README.md

[![Travis-CI Build
Status](https://travis-ci.org/hrbrmstr/AnomalyDetection.svg?branch=master)](https://travis-ci.org/hrbrmstr/AnomalyDetection)

<!-- README.md is generated from README.Rmd. Please edit that file -->

# AnomalyDetection

Anomaly Detection Using Seasonal Hybrid Extreme Studentized Deviate Test

## Description

A technique for detecting anomalies in seasonal univariate time series.
The methods uses are robust, from a statistical standpoint, in the
presence of seasonality and an underlying trend. These methods can be
used in wide variety of contexts. For example, detecting anomalies in
system metrics after a new software release, user engagement post an
‘A/B’ test, or for problems in econometrics, financial engineering,
political and social sciences.

## About This Fork

Twitterfolks launched this package in 2014. Many coding and package
standards have changed. The package now conforms to CRAN standards.

The plots were nice and all but terribly unnecessary. The two core
functions have been modified to only return tidy data frames (tibbles,
actually). This makes it easier to chain them without having to deal
with list element dereferencing.

Shorter, snake-case aliases have also been provided:

  - `ad_ts` for `AnomalyDetectionTs`
  - `ad_vec` for `AnomalyDetectionVec`

The original names are still in the package but the `README` and
examples all use the newer, shorter versions.

The following outstanding PRs from the original repo are included:

  - Added in PR
    [\#98](https://github.com/twitter/AnomalyDetection/pull/98/)
    (@gggodhwani)
  - Added in PR
    [\#93](%3Chttps://github.com/twitter/AnomalyDetection/pull/93)
    (@nujnimka)
  - Added in PR
    [\#69](%3Chttps://github.com/twitter/AnomalyDetection/pull/69)
    (@randakar)
  - Added in PR
    [\#44](%3Chttps://github.com/twitter/AnomalyDetection/pull/44)
    (@nicolasmiller)
  - PR [\#92](https://github.com/twitter/AnomalyDetection/pull/92)
    (@caijun) inherently resolved

If those authors find this repo, please add yourselves to the
`DESCRIPTION` as contirbutors.

## What’s Inside The Tin

The following functions are implemented:

  - `ad_ts`: Anomaly Detection Using Seasonal Hybrid ESD Test
  - `ad_vec`: Anomaly Detection Using Seasonal Hybrid ESD Test

## How the package works

The underlying algorithm – referred to as Seasonal Hybrid ESD (S-H-ESD)
builds upon the Generalized ESD test for detecting anomalies. Note that
S-H-ESD can be used to detect both global as well as local anomalies.
This is achieved by employing time series decomposition and using robust
statistical metrics, viz., median together with ESD. In addition, for
long time series (say, 6 months of minutely data), the algorithm employs
piecewise approximation - this is rooted to the fact that trend
extraction in the presence of anomalies in non-trivial - for anomaly
detection.

Besides time series, the package can also be used to detect anomalies in
a vector of numerical values. We have found this very useful as many
times the corresponding timestamps are not available. The package
provides rich visualization support. The user can specify the direction
of anomalies, the window of interest (such as last day, last hour),
enable/disable piecewise approximation; additionally, the x- and y-axis
are annotated in a way to assist visual data analysis.

## Installation

You can install AnomalyDetection from github with:

``` r
# install.packages("devtools")
devtools::install_github("hrbrmstr/AnomalyDetection")
```

## How to get started

``` r
library(AnomalyDetection)
library(hrbrthemes)
library(tidyverse)
```

``` r
data(raw_data)

res <- ad_ts(raw_data, max_anoms=0.02, direction='both')

glimpse(res)
## Observations: 131
## Variables: 2
## $ timestamp <dttm> 1980-09-25 16:05:00, 1980-09-29 06:40:00, 1980-09-29 21:44:00, 1980-09-30 17:46:00, 1980-09-30 1...
## $ anoms     <dbl> 21.3510, 193.1036, 148.1740, 52.7478, 49.6582, 35.6067, 32.5045, 30.0555, 31.2614, 30.2551, 27.38...

# for ggplot2
raw_data$timestamp <- as.POSIXct(raw_data$timestamp)

ggplot() +
  geom_line(
    data=raw_data, aes(timestamp, count), 
    size=0.125, color="lightslategray"
  )  +
  geom_point(
    data=res, aes(timestamp, anoms), color="#cb181d", alpha=1/3
  ) +
  scale_x_datetime(date_labels="%b\n%Y") +
  scale_y_comma() +
  theme_ipsum_rc(grid="XY")
```

<img src="README-unnamed-chunk-4-1.png" width="768" />

From the plot, we observe that the input time series experiences both
positive and negative anomalies. Furthermore, many of the anomalies in
the time series are local anomalies within the bounds of the time
series’ seasonality (hence, cannot be detected using the traditional
approaches). The anomalies detected using the proposed technique are
annotated on the plot. In case the timestamps for the plot above were
not available, anomaly detection could then carried out using the
AnomalyDetectionVec function; specifically, one can use the
`AnomalyDetectionVec()` method. The equivalent call to the above would
be:

``` r
ad_vec(raw_data[,2], max_anoms=0.02, period=1440, direction='both')
```

Often, anomaly detection is carried out on a periodic basis. For
instance, at times, one may be interested in determining whether there
was any anomaly yesterday. To this end, we support a flag only\_last
whereby one can subset the anomalies that occurred during the last day
or last hour.

``` r
data(raw_data)

res <- ad_ts(raw_data, max_anoms=0.02, direction='both', only_last="day")

glimpse(res)
## Observations: 25
## Variables: 2
## $ timestamp <dttm> 1980-10-05 01:12:00, 1980-10-05 01:13:00, 1980-10-05 01:14:00, 1980-10-05 01:15:00, 1980-10-05 0...
## $ anoms     <dbl> 56.4691, 54.9415, 52.0359, 47.7313, 50.5876, 48.2846, 44.6438, 42.3077, 38.8363, 41.0145, 39.5523...

# for ggplot2
raw_data$timestamp <- as.POSIXct(raw_data$timestamp)

ggplot() +
  geom_line(
    data=raw_data, aes(timestamp, count), 
    size=0.125, color="lightslategray"
  )  +
  geom_point(
    data=res, aes(timestamp, anoms), color="#cb181d", alpha=1/3
  ) +
  scale_x_datetime(date_labels="%b\n%Y") +
  scale_y_comma() +
  theme_ipsum_rc(grid="XY")
```

<img src="README-unnamed-chunk-6-1.png" width="768" />

Anomaly detection for long duration time series can be carried out by
setting the longterm argument to `TRUE`.

## Copyright & License

Copyright © 2015 Twitter, Inc. and other contributors

Licensed under the GPLv3

## Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](CONDUCT.md). By participating in this project you agree to
abide by its terms.


NEWS.md
# AnomalyDetection 2.0.1

* Swapped out `stringr` dependency for a `stringi` dependency since stringr also
  imports (and, hence, forces installation of) `glue` and `magrittr` which are not needed.
  `lubridate` still requires `stringr` but this is step one in the eventual removal
  of `lubridate`. The ultimate goal is to have this dependent only on `stringi`
  and `stats` and (possibly) even only just `stats`.
* Added package `prefix::` to all imported functions to improve package
  safety & readability.
* Added `stats` to the `Imports:` in `DESCRIPTION`.
* Changed tex formatting in comments to markdown and added `@md` roxgen tags
  where needed.
* Trimmed more long lines and reformatted a number of sections to improve
  readability.
* Added in URL references to the cited USENIX and Technometrics papers references.
* Cleaned up some status, warning and error messages
* Updated `Depends:` R version to require 3.0.0 given `lubridate`'s min R version.

# AnomalyDetection 2.0.0

* Added in PR <https://github.com/twitter/AnomalyDetection/pull/98/> (@gggodhwani)
* Added in PR <https://github.com/twitter/AnomalyDetection/pull/93> (@nujnimka)
* Added in PR <https://github.com/twitter/AnomalyDetection/pull/69> (@randakar)
* Added in PR <https://github.com/twitter/AnomalyDetection/pull/44 (@nicolasmiller)
* PR <https://github.com/twitter/AnomalyDetection/pull/92> (@caijun) inherently resolved
  since we return `POSIXct` objects in the data frames now
* Removed plotting code
* Removed tests that tested plotting code
* Updated tests since we only return data frames now
* Updated package to conform to modern CRAN standards
* Added a `NEWS.md` file to track changes to the package.


tests/testthat/test-ts.R
library(AnomalyDetection)
context("Evaluation: AnomalyDetectionTs")

test_that("last day, both directions", {
  data(raw_data)
  results <- AnomalyDetectionTs(raw_data, max_anoms=0.02, direction='both', only_last='day')
  expect_equal(length(results), 2)
  expect_equal(length(results[[2L]]), 25)
})

test_that("both directions, e_value, with longterm", {
  data(raw_data)
  results <- AnomalyDetectionTs(raw_data, max_anoms=0.02, direction='both', longterm=TRUE, e_value=TRUE)
  expect_equal(length(results), 3)
  expect_equal(length(results[[2L]]), 131)
})

test_that("both directions, e_value, threshold set to med_max", {
  data(raw_data)
  results <- AnomalyDetectionTs(raw_data, max_anoms=0.02, direction='both', threshold="med_max", e_value=TRUE)
  expect_equal(length(results), 3)
  expect_equal(length(results[[2L]]), 4)
})


tests/testthat/test-NAs.R
library(AnomalyDetection)

context("Testing NAs")

test_that("check handling of datasets with leading and trailing NAs", {
  data(raw_data)
  raw_data[1:10, "count"] <- NA
  raw_data[length(raw_data[[2L]]), "count"] <- NA
  results <- AnomalyDetectionTs(raw_data, max_anoms=0.02, direction='both')
  expect_equal(length(results), 2)
  expect_equal(length(results[[2L]]), 131)
})

test_that("check handling of datasets with NAs in the middle", {
  data(raw_data)
  raw_data[floor(length(raw_data[[2L]])/2), "count"] <- NA
  expect_error(AnomalyDetectionTs(raw_data, max_anoms=0.02, direction='both'))
})

test_that("check handling of datasets with leading and trailing NAs", {
  data(raw_data)
  raw_data[1:10, "count"] <- NA
  raw_data[length(raw_data[[2L]]), "count"] <- NA
  results <- AnomalyDetectionVec(raw_data[[2]], max_anoms=0.02, period=1440, direction='both')
  expect_equal(length(results), 2)
  expect_equal(length(results[[2L]]), 131)
})

test_that("check handling of datasets with NAs in the middle", {
  data(raw_data)
  raw_data[floor(length(raw_data[[2L]])/2), "count"] <- NA
  expect_error(AnomalyDetectionVec(raw_data[[2L]], max_anoms=0.02,
                                   period=1440, direction='both'))
})


tests/testthat/test-vec.R
library(AnomalyDetection)

context("Evaluation: AnomalyDetectionVec")

test_that("last period, both directions", {
  data(raw_data)
  results <- AnomalyDetectionVec(raw_data[[2L]], max_anoms=0.02, direction='both', period=1440, only_last=TRUE)
  expect_equal(length(results), 2)
  expect_equal(length(results[[2L]]), 25)
})

test_that("both directions, e_value, with longterm", {
  data(raw_data)
  results <- AnomalyDetectionVec(raw_data[[2L]], max_anoms=0.02, direction='both', period=1440, longterm_period=1440*14, e_value=TRUE)
  expect_equal(length(results), 3)
  expect_equal(length(results[[2L]]), 131)
})

test_that("both directions, e_value, threshold set to med_max", {
  data(raw_data)
  results <- AnomalyDetectionVec(raw_data[[2L]], max_anoms=0.02, direction='both', period=1440, threshold="med_max", e_value=TRUE)
  expect_equal(length(results), 3)
  expect_equal(length(results[[2L]]), 6)
})


R/date_utils.R
format_timestamp <- function(indf, index = 1) {
  if (class(indf[[index]])[1] == "POSIXlt") {
    return(indf)
  }
  if (stringr::str_detect(indf[[index]][1], "^\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2} \\+\\d{4}$")) {
    indf[[index]] <- strptime(indf[[index]], format = "%Y-%m-%d %H:%M:%S", tz = "UTC")
  }
  else if (stringr::str_detect(indf[[index]][1], "^\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2}$")) {
    indf[[index]] <- strptime(indf[[index]], format = "%Y-%m-%d %H:%M:%S", tz = "UTC")
  }
  else if (stringr::str_detect(indf[[index]][1], "^\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}$")) {
    indf[[index]] <- strptime(indf[[index]], format = "%Y-%m-%d %H:%M", tz = "UTC")
  }
  else if (stringr::str_detect(indf[[index]][1], "^\\d{2}/\\d{2}/\\d{2}$")) {
    indf[[index]] <- strptime(indf[[index]], format = "%m/%d/%y", tz = "UTC")
  }
  else if (stringr::str_detect(indf[[index]][1], "^\\d{2}/\\d{2}/\\d{4}$")) {
    indf[[index]] <- strptime(indf[[index]], format = "%m/%d/%Y", tz = "UTC")
  }
  else if (stringr::str_detect(indf[[index]][1], "^\\d{4}\\d{2}\\d{2}$")) {
    indf[[index]] <- strptime(indf[[index]], format = "%Y%m%d", tz = "UTC")
  }
  else if (stringr::str_detect(indf[[index]][1], "^\\d{4}/\\d{2}/\\d{2}/\\d{2}$")) {
    indf[[index]] <- strptime(indf[[index]], format = "%Y/%m/%d/%H", tz = "UTC")
  }
  else if (stringr::str_detect(indf[[index]][1], "^\\d{10}$")) {
    # Handle Unix seconds in milliseconds
    indf[[index]] <- as.POSIXlt(indf[[index]], origin = "1970-01-01", tz = "UTC")
  }

  return(indf)
}

get_gran <- function(tsdf, index=1) {
  n <- length(tsdf[[index]])
  # We calculate the granularity from the time difference between the last 2 entries (sorted)
  gran <- round(difftime(max(tsdf[[index]]), sort(tsdf[[index]], partial = n - 1)[n - 1],
    units = "secs"
  ))

  if (gran >= 86400) {
    return("day")
  }
  else if (gran >= 3600) {
    return("hr")
  }
  else if (gran >= 60) {
    return("min")
  }
  else if (gran >= 1) {
    return("sec")
  }
  else {
    return("ms")
  }
}


R/detect_anoms.R
# Detects anomalies in a time series using S-H-ESD.
#
# Args:
# 	 data: Time series to perform anomaly detection on.
# 	 k: Maximum number of anomalies that S-H-ESD will detect as a percentage of the data.
# 	 alpha: The level of statistical significance with which to accept or reject anomalies.
# 	 num_obs_per_period: Defines the number of observations in a single period,
#                        and used during seasonal decomposition.
# 	 use_decomp: Use seasonal decomposition during anomaly detection.
# 	 use_esd: Uses regular ESD instead of hybrid-ESD. Note hybrid-ESD is more
#             statistically robust.
# 	 one_tail: If TRUE only positive or negative going anomalies are detected
#              depending on if upper_tail is TRUE or FALSE.
# 	 upper_tail: If TRUE and one_tail is also TRUE, detect only positive going
#                (right-tailed) anomalies. If FALSE and one_tail is TRUE, only
#                detect negative (left-tailed) anomalies.
# 	 verbose: Additionally printing for debugging.
# Returns:
#   A list containing the anomalies (anoms) and decomposition components (stl).
detect_anoms <- function(data, k = 0.49, alpha = 0.05, num_obs_per_period = NULL,
                         use_decomp = TRUE, use_esd = FALSE, one_tail = TRUE,
                         upper_tail = TRUE, verbose = FALSE) {

  if (is.null(num_obs_per_period)) {
    stop("must supply period length for time series decomposition")
  }

  num_obs <- nrow(data)

  # Check to make sure we have at least two periods worth of data for anomaly context
  if (num_obs < num_obs_per_period * 2) {
    stop("Anom detection needs at least 2 periods worth of data")
  }

  # Check if our timestamps are posix
  posix_timestamp <- if (class(data[[1L]])[1L] == "POSIXlt") TRUE else FALSE

  # Handle NAs
  if (length(rle(is.na(c(NA, data[[2L]], NA)))$values) > 3) {
    stop(
      paste0(
        "Data contains non-leading NAs. We suggest replacing NAs with ",
        "interpolated values (see na.approx in Zoo package).",
        collapse = "")
    )
  } else {
    data <- stats::na.omit(data)
  }

  # -- Step 1: Decompose data. This returns a univarite remainder which will be
  # used for anomaly detection. Optionally, we might NOT decompose.
  stats::stl(
    stats::ts(data[[2L]], frequency = num_obs_per_period),
    s.window = "periodic",
    robust = TRUE
  ) -> data_decomp

  # Remove the seasonal component, and the median of the data to create the univariate remainder
  data.frame(
    timestamp = data[[1L]],
    count = (data[[2L]] - data_decomp$time.series[, "seasonal"] -
               stats::median(data[[2L]]))
  ) -> data

  # Store the smoothed seasonal component, plus the trend component for use in
  # determining the "expected values" option
  data.frame(
    timestamp = data[[1L]],
    count = (as.numeric(trunc(data_decomp$time.series[, "trend"] +
                                data_decomp$time.series[, "seasonal"])))
  ) -> data_decomp

  if (posix_timestamp) data_decomp <- format_timestamp(data_decomp)

  # Maximum number of outliers that S-H-ESD can detect (e.g. 49% of data)
  max_outliers <- trunc(num_obs * k)

  if (max_outliers == 0) {
    stop(paste0(
      "With longterm=TRUE, AnomalyDetection splits the data into 2 week periods by default. You have ",
      num_obs,
      " observations in a period, which is too few. Set a higher piecewise_median_period_weeks.")
    )
  }

  func_ma <- match.fun(stats::median)
  func_sigma <- match.fun(stats::mad)

  ## Define values and vectors.
  n <- length(data[[2L]])
  if (posix_timestamp) {
    R_idx <- as.POSIXlt(data[[1L]][1L:max_outliers], tz = "UTC")
  } else {
    R_idx <- 1L:max_outliers
  }

  num_anoms <- 0L

  # Compute test statistic until r=max_outliers values have been
  # removed from the sample.
  for (i in 1L:max_outliers) {

    if (verbose) message(paste(i, "/", max_outliers, "completed"))

    if (one_tail) {
      if (upper_tail) {
        ares <- data[[2L]] - func_ma(data[[2L]])
      } else {
        ares <- func_ma(data[[2L]]) - data[[2L]]
      }
    } else {
      ares <- abs(data[[2L]] - func_ma(data[[2L]]))
    }

    # protect against constant time series
    data_sigma <- func_sigma(data[[2L]])
    if (data_sigma == 0) break

    ares <- ares / data_sigma
    R <- max(ares)

    temp_max_idx <- which(ares == R)[1L]

    R_idx[i] <- data[[1L]][temp_max_idx]

    data <- data[-which(data[[1L]] == R_idx[i]), ]

    ## Compute critical value.
    if (one_tail) {
      p <- 1 - alpha / (n - i + 1)
    } else {
      p <- 1 - alpha / (2 * (n - i + 1))
    }

    t <- stats::qt(p, (n - i - 1L))
    lam <- t * (n - i) / sqrt((n - i - 1 + t**2) * (n - i + 1))

    if (R > lam) num_anoms <- i

  }

  if (num_anoms > 0) {
    R_idx <- R_idx[1L:num_anoms]
  } else {
    R_idx <- NULL
  }

  return(list(anoms = R_idx, stl = data_decomp))

}


R/AnomalyDetection-package.R
#' Anomaly Detection Using Seasonal Hybrid Extreme Studentized Deviate Test
#'
#' A technique for detecting anomalies in seasonal univariate time series.
#' The methods uses are robust, from a statistical standpoint, in the presence of
#' seasonality and an underlying trend. These methods can be used in
#' wide variety of contexts. For example, detecting anomalies in system metrics after
#' a new software release, user engagement post an 'A/B' test, or for problems in
#' econometrics, financial engineering, political and social sciences.
#'
#' @name AnomalyDetection
#' @docType package
#' @author Owen S. Vallis, Jordan Hochenbaum, Arun Kejariwal; Modernization
#'         contributions by Bob Rudis
#' @importFrom stats aggregate mad median na.omit qt quantile stl ts
#' @importFrom stringi stri_detect_regex
#' @importFrom lubridate days floor_date hours
NULL



R/ts_anom_detection.R
#' Anomaly Detection Using Seasonal Hybrid ESD Test
#'
#' A technique for detecting anomalies in seasonal univariate time series where the input is a
#' series of <timestamp, count> pairs.
#'
#' @md
#' @name AnomalyDetectionTs
#' @param x Time series as a two column data frame where the first column consists of the
#'        timestamps and the second column consists of the observations.
#' @param max_anoms Maximum number of anomalies that S-H-ESD will detect as a percentage of the
#'        data.
#' @param direction Directionality of the anomalies to be detected. One of:
#'        `pos`, `neg`, `both`.
#' @param alpha The level of statistical significance with which to accept or reject anomalies.
#' @param only_last Find and report anomalies only within the last day or hr in the time seriess.
#'        One of `NULL`, `day`, `hr`.
#' @param threshold Only report positive going anoms above the threshold specified. One of:
#'        `None`, `med_max`, `p95`, `p99`.
#' @param e_value Add an additional column to the anoms output containing the expected value.
#' @param longterm Increase anom detection efficacy for time series that are greater than a month.
#'        See `Details`` below.
#' @param piecewise_median_period_weeks The piecewise median time window as described in Vallis,
#'        Hochenbaum, and Kejariwal (2014). Defaults to 2.
#' @param verbose Enable debug messages
#' @param na.rm Remove any NAs in timestamps.(default: `FALSE`)
#' @return The returned value is a data frame containing timestamps, values,
#'         and optionally expected values.
#' @references
#' - Vallis, O., Hochenbaum, J. and Kejariwal, A., (2014)
#'   "A Novel Technique for Long-Term Anomaly Detection in the Cloud", 6th USENIX, Philadelphia, PA.
#'   (<https://www.usenix.org/system/files/conference/hotcloud14/hotcloud14-vallis.pdf>)
#'  - Rosner, B., (May 1983), "Percentage Points for a Generalized ESD Many-Outlier
#'    Procedure", Technometrics, 25(2), pp. 165-172. (<https://www.jstor.org/stable/1268549>)
#' @examples
#' data(raw_data)
#'
#' ad_ts(raw_data, max_anoms=0.02, direction='both')
#'
#' # To detect only the anomalies on the last day, run the following:
#'
#' ad_ts(raw_data, max_anoms=0.02, direction='both', only_last="day")
#' @seealso [ad_vec()]
#' @export
AnomalyDetectionTs <- function(x, max_anoms = 0.10, direction = "pos",
                               alpha = 0.05, only_last = NULL, threshold = "None",
                               e_value = FALSE, longterm = FALSE,
                               piecewise_median_period_weeks = 2,
                               verbose = FALSE, na.rm = FALSE) {

  # Check for supported inputs types
  if (!is.data.frame(x)) {
    stop("data must be a single data frame.")
  } else {
    if (ncol(x) != 2 || !is.numeric(x[[2]])) {
      stop(paste0("data must be a 2 column data.frame, with the first column being ",
                  "a set of timestamps, and the second coloumn being numeric values.",
                  collapse=""))
    }
    # Format timestamps if necessary
    if (!(class(x[[1]])[1] == "POSIXlt")) x <- format_timestamp(x)
  }

  # Rename data frame columns if necessary
  if (any((names(x) == c("timestamp", "count")) == FALSE)) {
    colnames(x) <- c("timestamp", "count")
  }

  if (!is.logical(na.rm)) stop("na.rm must be either TRUE or FALSE")

  # Deal with NAs in timestamps
  if (any(is.na(x$timestamp))) {
    if (na.rm) {
      x <- x[-which(is.na(x$timestamp)), ]
    } else {
      stop("timestamp contains NAs, please set na.rm to TRUE or remove the NAs manually.")
    }
  }

  # Sanity check all input parameters
  if (max_anoms > .49) {
    stop(paste("max_anoms must be less than 50% of the data points (max_anoms =",
               round(max_anoms * length(x[[2]]), 0), " data_points =", length(x[[2]]), ")."))
  } else if (max_anoms < 0) {
    stop("max_anoms must be positive.")
  } else if (max_anoms == 0) {
    warning("0 max_anoms results in max_outliers being 0.")
  }

  if (!direction %in% c("pos", "neg", "both")) {
    stop("direction options are: pos | neg | both.")
  }

  if (!(0.01 <= alpha || alpha <= 0.1)) {
    if (verbose) message("Warning: alpha is the statistical signifigance, and is usually between 0.01 and 0.1")
  }

  if (!is.null(only_last) && !only_last %in% c("day", "hr")) {
    stop("only_last must be either 'day' or 'hr'")
  }

  if (!threshold %in% c("None", "med_max", "p95", "p99")) {
    stop("threshold options are: None | med_max | p95 | p99.")
  }

  if (!is.logical(e_value)) {
    stop("e_value must be either TRUE or FALSE")
  }

  if (!is.logical(longterm)) {
    stop("longterm must be either TRUE or FALSE")
  }

  if (piecewise_median_period_weeks < 2) {
    stop("piecewise_median_period_weeks must be at greater than 2 weeks")
  }

  # -- Main analysis: Perform S-H-ESD

  # Derive number of observations in a single day.
  # Although we derive this in S-H-ESD, we also need it to be minutley later on so we do it here first.
  gran <- get_gran(x, 1)

  if (gran == "day") {
    num_days_per_line <- 7
    if (is.character(only_last) && only_last == "hr") {
      only_last <- "day"
    }
  } else {
    num_days_per_line <- 1
  }

  # Aggregate data to minutely if secondly
  if (gran == "sec" || gran == "ms") { # ref: https://github.com/twitter/AnomalyDetection/pull/69/files
    x <- format_timestamp(
      stats::aggregate(
        x[2],
        format(x[1], "%Y-%m-%d %H:%M:00"),
        sum)
    ) # ref: https://github.com/twitter/AnomalyDetection/pull/44
    gran <-  "min" # ref: https://github.com/twitter/AnomalyDetection/pull/98/files?diff=unified
  }

  period <- switch(
    gran,
    sec = 3600, # ref: https://github.com/twitter/AnomalyDetection/pull/93/files
    ms = 1000,  # ref: https://github.com/twitter/AnomalyDetection/pull/69/files
    min = 1440,
    hr = 24,
    # if the data is daily, then we need to bump the period to weekly to get multiple examples
    day = 7
  )
  num_obs <- length(x[[2]])

  if (max_anoms < 1 / num_obs) max_anoms <- 1 / num_obs

  # -- Setup for longterm time series

  # If longterm is enabled, break the data into subset data frames and store in all_data
  if (longterm) {
    # Pre-allocate list with size equal to the number of piecewise_median_period_weeks chunks in x + any left over chunk
    # handle edge cases for daily and single column data period lengths
    if (gran == "day") {
      # STL needs 2*period + 1 observations
      num_obs_in_period <- period * piecewise_median_period_weeks + 1
      num_days_in_period <- (7 * piecewise_median_period_weeks) + 1
    } else {
      num_obs_in_period <- period * 7 * piecewise_median_period_weeks
      num_days_in_period <- (7 * piecewise_median_period_weeks)
    }

    # Store last date in time series
    last_date <- x[[1]][num_obs]

    all_data <- vector(mode = "list", length = ceiling(length(x[[1]]) / (num_obs_in_period)))

    # Subset x into piecewise_median_period_weeks chunks
    for (j in seq(1, length(x[[1]]), by = num_obs_in_period)) {

      start_date <- x[[1]][j]
      end_date <- min(start_date + lubridate::days(num_days_in_period), x[[1]][length(x[[1]])])

      # if there is at least 14 days left, subset it, otherwise subset last_date - 14days
      if (difftime(end_date, start_date, units = "days") == as.difftime(num_days_in_period, units = "days")) {
        all_data[[ceiling(j / (num_obs_in_period))]] <- x[x[[1]] >= start_date & x[[1]] < end_date, ]
      } else {
        all_data[[ceiling(j / (num_obs_in_period))]] <-
          x[x[[1]] > (last_date - lubridate::days(num_days_in_period)) & x[[1]] <= last_date, ]
      }

    }

  } else {
    # If longterm is not enabled, then just overwrite all_data list with x as the only item
    all_data <- list(x)
  }

  # Create empty data frames to store all anoms and seasonal+trend component from decomposition
  all_anoms <- data.frame(timestamp = numeric(0), count = numeric(0))
  seasonal_plus_trend <- data.frame(timestamp = numeric(0), count = numeric(0))

  # Detect anomalies on all data (either entire data in one-pass, or in 2 week blocks if longterm=TRUE)
  for (i in 1:length(all_data)) {

    anomaly_direction <- switch(
      direction,
      "pos" = data.frame(one_tail = TRUE, upper_tail = TRUE), # upper-tail only (positive going anomalies)
      "neg" = data.frame(one_tail = TRUE, upper_tail = FALSE), # lower-tail only (negative going anomalies)
      "both" = data.frame(one_tail = FALSE, upper_tail = TRUE)
    ) # Both tails. Tail direction is not actually used.

    # detect_anoms actually performs the anomaly detection and returns the results in a list containing the anomalies
    # as well as the decomposed components of the time series for further analysis.
    s_h_esd_timestamps <- detect_anoms(all_data[[i]],
                                       k = max_anoms, alpha = alpha, num_obs_per_period = period,
                                       use_decomp = TRUE, use_esd = FALSE,
                                       one_tail = anomaly_direction$one_tail, upper_tail = anomaly_direction$upper_tail,
                                       verbose = verbose
    )

    # store decomposed components in local variable and overwrite s_h_esd_timestamps to contain only the anom timestamps
    data_decomp <- s_h_esd_timestamps$stl
    s_h_esd_timestamps <- s_h_esd_timestamps$anoms

    # -- Step 3: Use detected anomaly timestamps to extract the actual anomalies (timestamp and value) from the data
    if (!is.null(s_h_esd_timestamps)) {
      anoms <- all_data[[i]][(all_data[[i]][[1]] %in% s_h_esd_timestamps), ]
    } else {
      anoms <- data.frame(timestamp = numeric(0), count = numeric(0))
    }

    # Filter the anomalies using one of the thresholding functions if applicable
    if (threshold != "None") {

      # Calculate daily max values
      periodic_maxs <- tapply(x[[2]], as.Date(x[[1]]), FUN = max)

      # Calculate the threshold set by the user
      if (threshold == "med_max") {
        thresh <- stats::median(periodic_maxs)
      } else if (threshold == "p95") {
        thresh <- stats::quantile(periodic_maxs, .95)
      } else if (threshold == "p99") {
        thresh <- stats::quantile(periodic_maxs, .99)
      }
      # Remove any anoms below the threshold
      anoms <- anoms[anoms[[2]] >= thresh, ]

    }

    all_anoms <- rbind(all_anoms, anoms)
    seasonal_plus_trend <- rbind(seasonal_plus_trend, data_decomp)

  }

  # Cleanup potential duplicates
  all_anoms <- all_anoms[!duplicated(all_anoms[[1]]), ]
  seasonal_plus_trend <- seasonal_plus_trend[!duplicated(seasonal_plus_trend[[1]]), ]

  # -- If only_last was set by the user, create subset of the data that represent the most recent day
  if (!is.null(only_last)) {

    start_date <- x[[1]][num_obs] - lubridate::days(7)
    start_anoms <- x[[1]][num_obs] - lubridate::days(1)

    if (gran == "day") {
      # TODO: This might be better set up top at the gran check
      breaks <- 3 * 12
      num_days_per_line <- 7
    } else {
      if (only_last == "day") {
        breaks <- 12
      } else {
        # We need to change start_date and start_anoms for the hourly only_last option
        start_date <- lubridate::floor_date(x[[1]][num_obs] - lubridate::days(2), "day")
        start_anoms <- x[[1]][num_obs] - lubridate::hours(1)
        breaks <- 3
      }
    }

    # subset the last days worth of data
    x_subset_single_day <- x[x[[1]] > start_anoms, ]

    x_subset_week <- x[(x[[1]] <= start_anoms) & (x[[1]] > start_date), ]
    all_anoms <- all_anoms[all_anoms[[1]] >= x_subset_single_day[[1]][1], ]
    num_obs <- length(x_subset_single_day[[2]])

  }

  # Calculate number of anomalies as a percentage
  anom_pct <- (length(all_anoms[[2]]) / num_obs) * 100

  # If there are no anoms, then let's exit
  if (anom_pct == 0) {
    if (verbose) message("No anomalies detected.")
    return(data.frame())
  }

  # Fix to make sure date-time is correct and that we retain hms at midnight
  all_anoms[[1]] <- format(all_anoms[[1]], format = "%Y-%m-%d %H:%M:%S")

  # Store expected values if set by user
  if (e_value) {
    anoms <- data.frame(
      timestamp = all_anoms[[1]], anoms = all_anoms[[2]],
      expected_value = seasonal_plus_trend[[2]][as.POSIXlt(seasonal_plus_trend[[1]], tz = "UTC") %in% all_anoms[[1]]],
      stringsAsFactors = FALSE
    )
  } else {
    anoms <- data.frame(timestamp = all_anoms[[1]], anoms = all_anoms[[2]],
                        stringsAsFactors = FALSE)
  }

  # Make sure we're still a valid POSIXct datetime.
  # TODO: Make sure we keep original datetime format and timezone.

  anoms$timestamp <- as.POSIXct(anoms$timestamp, tz = "UTC")

  class(anoms) <- c("tbl_df", "tbl", "data.frame")

  return(anoms)
}

#' @rdname AnomalyDetectionTs
#' @export
ad_ts <- AnomalyDetectionTs